### ISSUE TITLE
Reduce the size of the stdlib

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#1542]
Part of https://github.com/pyodide/pyodide/issues/646#issuecomment-827735529

Also related to https://discuss.python.org/t/minifying-the-stdlib-in-pyodide/8414

Largest modules are (size is in kB),
```
$ du -a . | sort -n -r | head -n 20
11060   .
1732    ./encodings
746    ./distutils
664     ./pydoc_data
660     ./pydoc_data/topics.py
504     ./asyncio
436     ./email
336     ./xml
300     ./distutils/command
256     ./unittest
224     ./_pydecimal.py
208     ./http
176     ./importlib
172     ./logging
164     ./urllib
148     ./xml/dom
116     ./inspect.py
108     ./pydoc.py
```

**Encodings**

From https://github.com/pyodide/pyodide/issues/646#issuecomment-827524917 by @hoodmane 
> I still think it would be a good idea to figure out which text encodings are actually used. They take up a lot of space in stdlib. I took ~80% of them out on my build and have been using it this way for a while. I've only gotten one missing codec error so far.

Reading https://docs.python.org/3/library/codecs.html#standard-encodings maybe the ideal case would be by default to only include UTF-8, possibly latin1 (maybe some Python files still use it?) and possibly Python specific encodings. Then put all the rest into an independent optional package. 

**distutils**

Per [PEP 632](https://www.python.org/dev/peps/pep-0632/) is going to be deprecated in Python 3.10 and removed in 3.12, so we might as well unpackage it as a standalone package earlier. I'm not sure why we would need it at runtime.

**pydoc + pydoc_data**

Potentially we could make it standalone, but it's necessary to make "help" work and also used in a few places in stdlib.

**Other**

Then there is all the work by @phorward in https://github.com/pyodide/pyodide/issues/646#issuecomment-823657002 and suggestions from https://discuss.python.org/t/minifying-the-stdlib-in-pyodide/8414

**Comment by hoodmane** [comment]
> the ideal case would be by default to only include UTF-8, possibly latin1 (maybe some Python files still use it?) and possibly Python specific encodings. 

This is a good starting point, but I found that Python uses weird encodings more often than you would hope, so I suspect we'll need to throw a few odd ones back in. latin1 is definitely necessary.

**Comment by hoodmane** [comment]
> Potentially we could make it standalone, but it's necessary to make "help" work and also used in a few places in stdlib. So that's probably less obvious.

Maybe we could stub out `help` with a function telling you to load `pydoc` if you want to use help?

**Comment by phorward** [comment]
Hello @rth,
thanks for opening the discussion for feedback on [python.org](https://discuss.python.org/t/minifying-the-stdlib-in-pyodide/8414).

Blackward's proposal there, keeping the full stdlib with Pyodide but loading its modules on demand, seems to be a reasonable idea. On my [pyodide-nano](https://github.com/phorward/pyodide/tree/pyodide-nano) branch a lot of not essentially required modules [are removed](https://github.com/phorward/pyodide/blob/pyodide-nano/cpython/remove_modules.txt) are removed from the Pyodide package (and also stripped using [pyminify](https://github.com/phorward/pyodide/blob/pyodide-nano/cpython/Makefile#L45)). The modules removed there could be packaged or made available to the Pyodide build afterwards and loaded on demand.

Regarding `.pyc` files, I'm not sure if they really safe that much space. A gzip-compressed python-text-file in most cases gets a better compression rate than a .pyc binary file.

**Comment by rth** [comment]
> keeping the full stdlib with Pyodide but loading its modules on demand, seems to be a reasonable idea

yeah, I agree. At least for large and infrequently used modules. I started with distutils in https://github.com/pyodide/pyodide/pull/1543 . There is probably a cost-benefit analysis that can be done per module. As if we make too many modules optional (with an explicit load) it will be annoying to use in micropip installed packages that may have arbitrary imports. I guess we'll have to write a custom import hook to handle it eventually, but for the few modules mentioned here I was hoping we could do without it.

> A gzip-compressed python-text-file in most cases gets a better compression rate than a .pyc binary file.

True, but you still have to copy it uncompressed in the virtual memory. I'm not saying we should start shipping .pyc, I just mentioned it there for completeness.

**Comment by rth** [comment]
> On my pyodide-nano branch a lot of not essentially required modules are removed from the Pyodide package 

Just curious do you see any impact on the load time with that branch? I guess it might be hard to benchmark as the Python's `http.server` is not great for that.

**Comment by phorward** [comment]
> yeah, I agree. At least for large and infrequently used modules. I started with distutils in #1543 . There is probably a cost-benefit analysis that can be done per module.

OK, but then we need to find some kind of automatism, especially for changes in future Python versions. What about loading single files on demand, instead of creating a package for each, as demonstrated with #1543? This would satisfy for most modules of the stdlib.

> Just curious do you see any impact on the load time with that branch?

I didn't do any benchmarks on it yet, but I am sure that the startup time in generally will only be a little faster. I can do some benchmarks on master and my branch.

> Also what's the final size you get?

These a the sizes of my last build on this branch. It is a Pyodide binary that is able to run the UI applications we're working on, but I think it won't be able to run any modules from the Python scientifc stack.
```
3.3M    ./pyodide.asm.wasm
820K    ./pyodide.asm.data
304K    ./pyodide.asm.js
```

**Comment by rth** [comment]
> What about loading single files on demand, instead of creating a package for each

The issue is that wouldn't work with our dependency resolution and `loadPackage`, and we would need some other mechanism. Also the stdlib has 203 modules, I don't think we would want to unvendor them all as that would result in a large number of HTTP requests for common usage.

Also I think we won't be able to load required modules automatically (with import hooks), until we have a solution for https://github.com/pyodide/pyodide/issues/1503 since packages are loaded async.

**Comment by hoodmane** [comment]
Once we get #1503 implemented, in contexts where we have syncio we could add a blocking installer to the metapath. My demo implementation in #1545 allows scheduling a blocking task before blocking on it, though I haven't yet implemented a "syncio.gather" which we would want for dependency resolution.

**Comment by rth** [comment]
> I still think it would be a good idea to figure out which text encodings are actually used. 

So when loading the interpreter,
```
>>> import sys
>>> [el for el in list(sys.modules) if "enc" in el]
['encodings.aliases', 'encodings', 'encodings.utf_8', 'encodings.latin_1']
```

It would be useful to load some of the other packages in Pyodide to check if other encodings are loaded. But generally unvendoring the rest of the encodings to an standalone `encodings` package (as was done for https://github.com/pyodide/pyodide/pull/1543) would likely have a measurable size impact. cc @chrysn

**Comment by rth** [comment]
Here is a list of deprecated stlib modules and associated sizes,

| package name | size (kB)                            |
|--------------|--------------------------------------|
| aifc         | 32                                   |
| asynchat     | 12                                   |
| asyncore     | 20                                   |
| audioop      | audioop.so is 323 KB                 |
| cgi          | 34                                   |
| cgitb        | 12                                   |
| chunk        | 5.4                                  |
| crypt        | _crypt.so is 41K + 3.8K for crypt.py |
| imghdr       | 3.8                                  |
| mailcap      | 8                                    |
| msilib       | 110                                  |
| nis          | 75                                   |
| ossaudiodev  | 125                                  |
| pipes        | 8.8                                  |
| smtpd        | 35                                   |
| sndhdr       | 7                                    |
| spwd         | 60                                   |
| sunau        | 18                                   |
| telnetlib    | 23                                   |
| uu           | 6.8                                  |
| xdrlib       | 5.8                                  |

The .so sizes are indicated for x86_64, so they might be different for WASM.

I think what we should do is to add a `stdlib-deprecated` package and unvendor those there. The problem of removing  them altogether is some people still use them, for instance, audioop was added after a user request in https://github.com/pyodide/pyodide/issues/801

The only thing I'm not sure is how to dynamically link some C extension modules in CPython stdlib, and statically link the rest, but that's likely solvable. We can probably start by already unvendoring all those pure python modules.
